src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java

```diff
@@ -0,0 +1,229 @@
+      throws IOException, OAuth2Exception {
+    List<String> info =
+        verifyAndGetInfoFromJwt(
+            oauth2Request.getRequestBody().getAssertion(), new JwtSigningKeyResolver(keyUrl));
+    String email = info.get(0);
+    String googleAccountId = info.get(1);
+    switch (oauth2Request.getRequestBody().getIntent()) {
+      case CHECK:
+        handleCheckAssertion(response, email, googleAccountId);
+        break;
+      case GET:
+        handleGetAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      case CREATE:
+        handleCreateAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
+      throws OAuth2Exception {
+    Jws<Claims> jws;
+
+    String email = null;
+    String googleAccountId = null;
+    try {
+      jws = Jwts.parserBuilder().setSigningKeyResolver(keys).build().parseClaimsJws(assertion);
+      if (!jws.getBody().getIssuer().equals("https://accounts.google.com")) {
+        throw new InvalidRequestException(ErrorCode.INVALID_JWT_ISS);
+      }
+      if (!jws.getBody().getAudience().equals(aud)) {
+        throw new InvalidRequestException(ErrorCode.WRONG_JWT_AUD);
+      }
+      email = jws.getBody().get("email", String.class);
+      googleAccountId = jws.getBody().get("sub", Long.class).toString();
+    } catch (JwtException ex) {
+      log.log(Level.INFO, "JWT Decode ERROR!", ex);
+      throw new InvalidRequestException(ErrorCode.INVALID_JWT);
+    }
+    return ImmutableList.of(email, googleAccountId);
+  }
+
+  public void handleCheckAssertion(
+      HttpServletResponse response, String email, String googleAccountId) throws IOException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    JSONObject json = new JSONObject();
+    if (user.isPresent()) {
+      json.appendField("account_found", "true");
+      response.setContentType("application/json;charset=UTF-8");
+      response.getWriter().println(json.toJSONString());
+    } else {
+      json.appendField("account_found", "false");
+      response.setContentType("application/json;charset=UTF-8");
+      response.getWriter().println(json.toJSONString());
+      response.setStatus(HttpStatus.SC_NOT_FOUND);
+    }
+    response.getWriter().flush();
+  }
+
+  public void handleGetAssertion(
+      HttpServletResponse response, String email, String googleAccountId, List<String> scopes)
+      throws IOException, InvalidScopeException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    if (user.isPresent()) {
+      ClientDetails client = clientDetailsService.getClientByID("google").get();
```

Comment:
Shouldn't the client id be passed in and verified with client secret? Same for other place

Reply:
Actually from the description of Assertion, I do not see client id and secret.
I let the filter set google's client for it.
I can read the client from session here but they are actually the same.

Reply:
It's not in the assertion content, but either in the header or body.

Reply:
But in https://developers.google.com/identity/account-linking/oauth-with-sign-in-linking?oauth=code,
I do not see that. The sample is:
POST /token HTTP/1.1
Host: oauth2.example.com
Content-Type: application/x-www-form-urlencoded

grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&intent=check&assertion=JWT&scope=SCOPES

Reply:
For your case, it should be in post parameter.